Migration tutorials to v1.1.1 (#138)
casper-ecosystem/enhanced-nft: tutorials/custom-migration-tutorial.md

# Upgrading to v1.1.1 using a Custom NamedKey Convention

This tutorial uses the Casper command-line client to upgrade *and* migrate from an NFT contract installed using release [v1.0.0](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/tag/v1.0.0) and customized NamedKeys. The outcome of the upgrade will be a [v1.1.1](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/tag/v1.1.1) NFT contract with its storage structure migrated to the new format.

## Prerequisites

> - You have previously installed a CEP-78 contract using release v1.0.0 on a Casper network. Thus, you have experience with: the [Casper CEP-78 NFT Standard](https://github.com/casper-ecosystem/cep-78-enhanced-nft/), the Casper command-line client, and interacting with a Casper network.

Reviewer: `with:` reads a bit strange

Author: Thanks for catching this, I used to have bullets. I removed the `:` .

Author: Thanks, I have removed the `:`. It used to be a list.

- `node-address`: An IP address of a peer on the network. The default port for JSON-RPC servers on Mainnet and Testnet is 7777.
- `chain-name`: The chain name of the network where you wish to send the deploy. For Mainnet, use *casper*. For Testnet, use *casper-test*.
- `secret-key`: The file name containing the secret key of the account paying for the deploy.
- `payment-amount`: The payment for the deploy in motes.
- `session-path`: The path to the compiled Wasm on your computer. When using the [cep-78-wasm.tar.gz](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/download/v1.1.1/cep-78-wasm.tar.gz) provided, this would be the path to the `contract.wasm` file.
- `named_key_convention`: Argument that specifies the use of the `V_1_0_standard` [NamedKeyConvention](../README.md#namedkeyconventionmode).
- `collection_name`: Argument that specifies the collection name as a String.
- `hash_key_name`: The custom contract package hash NamedKey as a String.
- `access_key_name`: The custom contract package access NamedKey as a String.

The command returns the deploy hash that you can use to verify whether or not the deploy succeeded.

**Important Notes**: 

> - Suppose an account attempts to install a second CEP-78 contract instance with the same collection name, but a custom contract package access NamedKey is used. In this case, the installation will not overwrite the NamedKey entry under which the access URef is written.
> - When installing a new version of the CEP-78 contract, you do not need to specify all the runtime arguments needed during the initial installation of version 1.0.0, such as `total_token_supply`, `ownership_mode`, etc. 

Reviewer: ```suggestion
- Suppose an account attempts to upgrade by installing a second CEP-78 contract instance with the same collection name, but a custom contract package access NamedKey is used. In this case, the installation will not overwrite the NamedKey entry under which the access URef is written.
- When upgrading by installing a second a CEP-78 contract, you do not need to specify all the runtime arguments needed during the initial installation of version 1.0.0, such as `total_token_supply`, `ownership_mode`, etc. 
```

Author: I should remove the first point because it is confusing. I wanted to explain why custom NamedKeys are used. But it isn't another upgrade path. If I add "upgrade" in there, it will get confusing. Whoever is on this tutorial knows why they used custom NamedKeys.

I will keep the second point and suggestion by saying: "When upgrading by installing a new version of the CEP-78 contract...".  Because it might be a 3rd upgrade (v1.0.0, v1.1.0, v1.1.1).

Author: I have removed the first point from the standard path as well, to avoid confusion. The readme is already clear enough.

**Example command to upgrade to v1.1.1:**

The following is an example of upgrading and migrating to version 1.1.1 of a previously installed NFT collection using version 1.0.0 and with custom NamedKeys.

```bash
casper-client put-deploy \
--node-addres http://65.21.235.219:7777 \
--chain-name "casper-test" \
--secret-key ~/KEYS/secret_key.pem \
--payment-amount 300000000000 \
--session-path contract.wasm \
--session-arg "named_key_convention:u8='2'" \
--session-arg "collection_name:string='CEP-78-collection'" \
--session-arg "hash_key_name:string='mangled_hash_key'" \
--session-arg "access_key_name:string='mangled_access_key'"
```

**Other Examples:**

- A [Testnet account](https://testnet.cspr.live/account/013060d19fa5d5e471c2bbe88f35871454d2e162c444100eaca34671339c78ced4) that uses the custom NamedKeys "mangled_hash_key" and "mangled_access_key".

> - An example [Testnet deploy](https://testnet.cspr.live/deploy/59a785471600e183718b790fb19b3dec7242fde105928b9f90f01347b3c65f46) that specified the custom NamedKey convention.

Reviewer: If I follow this link named_key_convention used was 1, not 2. It's a bit confusing.

Author: Thanks for catching this, I will updated it to https://testnet.cspr.live/deploy/55cb135c9b600263baedf72b124b02ff6dd74dd27d2d9be444b1bed6ee5e3301.

# Upgrading to v1.1.1 using a Custom NamedKey Convention

This tutorial uses the Casper command-line client to upgrade *and* migrate from an NFT contract installed using release [v1.0.0](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/tag/v1.0.0) and customized NamedKeys. The outcome of the upgrade will be a [v1.1.1](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/tag/v1.1.1) NFT contract with its storage structure migrated to the new format.

## Prerequisites

- You have previously installed a CEP-78 contract using release v1.0.0 on a Casper network. Thus, you have experience with the [Casper CEP-78 NFT Standard](https://github.com/casper-ecosystem/cep-78-enhanced-nft/), the Casper command-line client, and interacting with a Casper network.

> - The v1.0.0 NFT contract uses custom NamedKeys for the contract package hash and contract package access URef. An example of how the NamedKeys could have been customized is found [here](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/test-contracts/mangle_named_keys/src/main.rs).

Reviewer 2: ```suggestion
- Your v1.0.0 NFT contract instance uses custom NamedKeys for the contract package hash and contract package access URef. An example of how the NamedKeys could have been customized is found [here](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/test-contracts/mangle_named_keys/src/main.rs).
```

As is, it sounds like the overall v1.0.0 NFT contract uses custom NamedKeys as standard.

- You understand what is [New in Version 1.1](https://github.com/casper-ecosystem/cep-78-enhanced-nft/#new-in-version-11) of the CEP-78 Enhanced NFT Standard.

## Upgrading and Migrating Terminology

The upgrade to version 1.1.1 involves a data migration to a new [page system](https://github.com/casper-ecosystem/cep-78-enhanced-nft#the-cep-78-page-system) tracking token ownership. The usual [upgrade](https://docs.casperlabs.io/dapp-dev-guide/writing-contracts/upgrading-contracts/) process triggers the data migration. For more information, see [Standard Migration Tutorial](standard-migration-tutorial.md#upgrading-and-migrating-terminology).

## Steps to Upgrade to Version 1.1.1

Navigate to the [v1.1.1 release](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/tag/v1.1.1) and download the [cep-78-wasm.tar.gz](https://github.com/casper-ecosystem/cep-78-enhanced-nft/releases/download/v1.1.1/cep-78-wasm.tar.gz) file. Unarchive it to have access to the Wasm files provided. The Rust source code is also available if you would like to build and test the contract according to the Makefile provided. 

> The `cep-78-wasm` folder contains the `contract.wasm` to send to the network to upgrade the NFT contract to version 1.1.1. In addition, you will see other useful Wasm files to interact with the contract once it is stored on-chain.

Reviewer 3: Double check the folder name

Author: Done. It is correct once unzipped. 
<img width="277" alt="Screen Shot 2023-01-17 at 9 16 42 PM" src="https://user-images.githubusercontent.com/4185994/213002717-f6edd5b9-882d-4a27-b518-7d4a8f6932fe.png">

### Custom NamedKeys before Migration

The custom migration path assumes that the contract has modified the NamedKey entries created during the v1.0.0. See the example below as well as the [NamedKeyConvention](https://github.com/casper-ecosystem/cep-78-enhanced-nft#namedkeyconventionmode) modality. 

| NamedKey Pre-Migration | Explanation |
|-------------|-------------|
| nft_contract | The hash identifying the NFT contract |
| A user-specified String | The hash identifying the contract package containing the NFT contract. In this example, it is "mangled_hash_key" | 

> | A user-specified String | The URef used as an access token or reference to the contract package. In this example, it is "mangled_access_key" | 

Reviewer 3: ```suggestion
| A user-specified String | The URef used as an access token to the contract package. In this example, it is "mangled_access_key" | 
```
